### PR TITLE
SSV-Token-DEX-token-swap

### PR DESCRIPTION
**Pull request by vadiminc**

### CODE REVIEW

File: contracts/DEX.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+contract DEX {
+    event CDTToSSVConverted(uint256 amount);
+
+    IERC20 public cdtToken;
+    IERC20 public ssvToken;
+
+    uint public rate;
+
+    function init(address _cdtTokenAddress, address _ssvTokenAddress) public {
```
Comment (adamzigdon) [inline contracts/DEX.sol:13]: you can change the parameters to be of type `IERC20` and then you won't need to cast in the following lines.

File: contracts/DEX.sol [hunk 2]
```diff
@@ -0,0 +1,25 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract DEX {
+    event CDTToSSVConverted(uint256 amount);
+
+    IERC20 public cdtToken;
+    IERC20 public ssvToken;
```
Comment (adamzigdon) [inline contracts/DEX.sol:9]: if we want this contract to be generic you should rename the tokens variables to be `sourceToken` and `destinationToken`/`targetToken`.

File: contracts/DEX.sol [hunk 3]
```diff
@@ -0,0 +1,25 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract DEX {
+    event CDTToSSVConverted(uint256 amount);
```
Comment (adamzigdon) [inline contracts/DEX.sol:6]: we should think if an event is really necessary here.
if so, i would remove the token names here and make it more general.
also, we should add the user address and have both amounts (source and target).

File: contracts/DEX.sol [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    uint public rate;
+
+    function init(address _cdtTokenAddress, address _ssvTokenAddress) public {
+        cdtToken = IERC20(_cdtTokenAddress);
+        ssvToken = IERC20(_ssvTokenAddress);
+        rate = 10;
+    }
+
+    function convertCDTToSSV(uint256 amount) public {
```
Comment (adamzigdon) [inline contracts/DEX.sol:19]: rename this method as well.

File: contracts/SSVToken.sol [hunk 5]
```diff
@@ -0,0 +1,9 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract SSVToken is ERC20 {
```
Comment (adamzigdon) [inline contracts/SSVToken.sol:5]: this file shouldn't be here.
if you want to have a token for testing it you should create a `mocks` folder.

File: contracts/tmp/OLDToken.sol [hunk 6]
```diff
@@ -0,0 +1,9 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract OLDToken is ERC20 {
```
Comment (adamzigdon) [inline contracts/tmp/OLDToken.sol:5]: same here.

File: hardhat.config.ts [hunk 7]
```diff
@@ -40,7 +40,13 @@ const config = {
       }
     ],
   },
-  networks: {},
+  networks: {
+    ganashe: {
```
Comment (adamzigdon) [inline hardhat.config.ts:44]: this is misspelled (ganache).

File: hardhat.config.ts [hunk 8]
```diff
@@ -40,7 +40,13 @@ const config = {
...
+      url: 'http://127.0.0.1:7545',
+      accounts: [`0x15da52959efa15bc7a7b3c7bc293425e2e493a626a7c51e77b2b4326adcbe822`],
```
Comment (adamzigdon) [inline hardhat.config.ts:46]: that shouldn't be hardcoded (use env instead).

File: contracts/tmp/OLDToken.sol (hunk 6 above)
Comment (adamzigdon) [reply contracts/tmp/OLDToken.sol:5]: also why is `OLD` in capital letters?

File: hardhat.config.ts [hunk 9]
```diff
@@ -52,7 +58,7 @@ if (process.env.GOERLI_ETH_NODE_URL) {
   config.networks['goerli'] = {
     url: process.env.GOERLI_ETH_NODE_URL,
     accounts: [`0x${process.env.GOERLI_OWNER_PRIVATE_KEY}`],
-    gasPrice: 98000000000
+    gasPrice: 320000000000
```
Comment (adamzigdon) [inline hardhat.config.ts:61]: i think that we should have a way to overwrite it via env and make the default smaller (320 gwei is very high).

File: scripts/dex-deploy.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  const [deployer] = await ethers.getSigners();
+  console.log(
+    'Deploying the contracts with the account:',
+    await deployer.getAddress()
+  );
+
+  console.log('Account balance:', (await deployer.getBalance()).toString());
+
+  const Contract = await ethers.getContractFactory('DEX');
+  const contract = await upgrades.deployProxy(
```
Comment (adamzigdon) [inline scripts/dex-deploy.ts:14]: it's a very bad practice to have two variables that differ only in their case.

File: scripts/ssv-token-deploy.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import { ethers } from 'hardhat';
+
+async function main() {
```
Comment (adamzigdon) [inline scripts/ssv-token-deploy.ts:3]: mocks are used only for test purposes, they should be moved to test folder.

**Review by adamzigdon — CHANGES_REQUESTED** [review]
see comments

File: contracts/DEX.sol (hunk 2 above)
Comment (vadiminc) [reply contracts/DEX.sol:10]: i think no, Lior wants to have two ways swap

File: contracts/DEX.sol (hunk 1 above)
Comment (vadiminc) [reply contracts/DEX.sol]: done

File: contracts/DEX.sol [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+    IERC20 public cdtToken;
+    IERC20 public ssvToken;
+
+    uint public rate;
+
+    function init(address _cdtTokenAddress, address _ssvTokenAddress) public {
+        cdtToken = IERC20(_cdtTokenAddress);
+        ssvToken = IERC20(_ssvTokenAddress);
+        rate = 10;
```
[the remark this answers is not in the document]
Comment (vadiminc) [reply contracts/DEX.sol]: done

File: contracts/DEX.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (vadiminc) [reply contracts/DEX.sol]: done

File: contracts/DEX.sol (hunk 3 above)
Comment (vadiminc) [reply contracts/DEX.sol:6]: will need to discuss with Lior currently just for 2 our tokens

File: contracts/SSVToken.sol (hunk 5 above)
Comment (vadiminc) [reply contracts/SSVToken.sol]: moved

File: contracts/tmp/OLDToken.sol (hunk 6 above)
Comment (vadiminc) [reply contracts/tmp/OLDToken.sol]: moved, just example

File: hardhat.config.ts (hunk 7 above)
Comment (vadiminc) [reply hardhat.config.ts]: changed

File: hardhat.config.ts (hunk 8 above)
Comment (vadiminc) [reply hardhat.config.ts]: fixed

File: hardhat.config.ts (hunk 9 above)
Comment (vadiminc) [reply hardhat.config.ts]: ok

File: scripts/dex-deploy.ts (hunk 10 above)
Comment (vadiminc) [reply scripts/dex-deploy.ts:14]: agreed, changed

File: scripts/ssv-token-deploy.ts (hunk 11 above)
Comment (vadiminc) [reply scripts/ssv-token-deploy.ts]: done

File: contracts/DEX.sol (hunk 3 above)
Comment (lior-blox) [reply contracts/DEX.sol:6]: Not sure we need an event

**Pull request merged**